File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/LoinExportExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.model.SimpleLoinPublicDto;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import com.pb40.bimportal.util.ExportUtils;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * LOIN export examples for BIM Portal Java client.
 *
 * <p>Demonstrates LOIN export workflows in multiple formats including PDF, OpenOffice, OKSTRA,
 * LOIN-XML, and IDS with automatic content type detection.
 */
public class LoinExportExample {

  private static final Logger logger = LoggerFactory.getLogger(LoinExportExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Detect content type from byte array and return appropriate file extension.
   *
   * @param content The byte array content
   * @param defaultExtension Default extension if detection fails
   * @return Detected file extension
   */
  private static String detectFileExtension(byte[] content, String defaultExtension) {
    if (content == null || content.length < 4) {
      return defaultExtension;
    }

    // Check for ZIP file signature (PK)
    if (content.length >= 2 && content[0] == 0x50 && content[1] == 0x4B) {
      logger.debug("Detected ZIP file signature");
      return "zip";
    }

    // Check for PDF signature (%PDF)
    if (content.length >= 4
        && content[0] == 0x25
        && content[1] == 0x50
        && content[2] == 0x44
        && content[3] == 0x46) {
      logger.debug("Detected PDF file signature");
      return "pdf";
    }

    // Check for OpenDocument format (ZIP-based, but starts differently)
    if (content.length >= 4 && content[0] == 0x50 && content[1] == 0x4B) {
      // Additional check for OpenDocument by looking for mimetype file
      // This is a more sophisticated check for ODT files
      String contentStr = new String(content, 0, Math.min(1024, content.length));
      if (contentStr.contains("mimetypeapplication/vnd.oasis.opendocument")) {
        logger.debug("Detected OpenDocument format");
        return "odt";
      }
      logger.debug("Detected ZIP-based format (possibly ODT or other)");
      return "zip";
    }

    // Check for XML content (IDS files are typically XML)
    if (content.length >= 5) {
      String start = new String(content, 0, Math.min(100, content.length));
      if (start.trim().startsWith("<?xml") || start.contains("<ids") || start.contains("<loin")) {
        logger.debug("Detected XML format");
        return "xml";
      }
    }

    logger.debug("Could not detect file type, using default: {}", defaultExtension);
    return defaultExtension;
  }

  /**
   * Export content with automatic file type detection.
   *
   * @param content The byte array content
   * @param baseFilename Base filename without extension
   * @param expectedExtension Expected file extension for logging
   * @return Optional path to saved file
   */
  private static Optional<Path> exportWithDetection(
      byte[] content, String baseFilename, String expectedExtension) {
    String detectedExtension = detectFileExtension(content, expectedExtension);
    String filename = baseFilename + "." + detectedExtension;

    if (!detectedExtension.equals(expectedExtension)) {
      logger.info(
          "Content type detection: expected '{}' but detected '{}' for {}",
          expectedExtension,
          detectedExtension,
          baseFilename);
    }

    return ExportUtils.saveExportFile(content, filename);
  }

  /**
   * Run LOIN export examples.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runLoinExportExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("📋 LOIN EXPORT EXAMPLES");
    System.out.println("============================================================");

    System.out.println("\n1️⃣ Searching for available LOINs...");
    try {
      List<SimpleLoinPublicDto> loins = client.searchLoins();
      if (loins.isEmpty()) {
        System.out.println("🔭 No LOINs found for export");
        return;
      }

      System.out.println("✅ Found " + loins.size() + " LOINs:");
      for (int i = 0; i < Math.min(3, loins.size()); i++) {
        SimpleLoinPublicDto loin = loins.get(i);
        System.out.println("   " + (i + 1) + ". " + loin.getName() + " (" + loin.getGuid() + ")");
      }

      SimpleLoinPublicDto selectedLoin = loins.get(0);
      System.out.println("\n🎯 Using LOIN: '" + selectedLoin.getName() + "'");

      Map<String, Path> exportResults = new HashMap<>();

      System.out.println("\n2️⃣ Exporting LOIN in multiple formats...");

      // PDF Export with auto-detection
      System.out.println("   📄 Exporting as PDF...");
      Optional<byte[]> pdfContent = client.exportLoinPdf(selectedLoin.getGuid());
      if (pdfContent.isPresent()) {
        String baseFilename = "loin_pdf_" + selectedLoin.getGuid();
        Optional<Path> pdfPath = exportWithDetection(pdfContent.get(), baseFilename, "pdf");
        if (pdfPath.isPresent()) {
          exportResults.put("PDF", pdfPath.get());
          System.out.println("   ✅ PDF exported: " + pdfPath.get());
        } else {
          System.out.println("   ❌ PDF export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ PDF export failed: No content received");
      }

      // OpenOffice Export with auto-detection
      System.out.println("   📝 Exporting as OpenOffice...");
      Optional<byte[]> odtContent = client.exportLoinOpenOffice(selectedLoin.getGuid());
      if (odtContent.isPresent()) {
        String baseFilename = "loin_odt_" + selectedLoin.getGuid();
        Optional<Path> odtPath = exportWithDetection(odtContent.get(), baseFilename, "odt");
        if (odtPath.isPresent()) {
          exportResults.put("OpenOffice", odtPath.get());
          System.out.println("   ✅ OpenOffice exported: " + odtPath.get());
        } else {
          System.out.println("   ❌ OpenOffice export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OpenOffice export failed: No content received");
      }

      // OKSTRA Export with auto-detection
      System.out.println("   🏗️ Exporting as OKSTRA...");
      Optional<byte[]> okstraContent = client.exportLoinOkstra(selectedLoin.getGuid());
      if (okstraContent.isPresent()) {
        String baseFilename = "loin_okstra_" + selectedLoin.getGuid();
        Optional<Path> okstraPath = exportWithDetection(okstraContent.get(), baseFilename, "zip");
        if (okstraPath.isPresent()) {
          exportResults.put("OKSTRA", okstraPath.get());
          System.out.println("   ✅ OKSTRA exported: " + okstraPath.get());
        } else {
          System.out.println("   ❌ OKSTRA export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OKSTRA export failed: No content received");
      }

      // LOIN-XML Export with auto-detection
      System.out.println("   🔗 Exporting as LOIN-XML...");
      Optional<byte[]> xmlContent = client.exportLoinXml(selectedLoin.getGuid());
      if (xmlContent.isPresent()) {
        String baseFilename = "loin_xml_" + selectedLoin.getGuid();
        Optional<Path> xmlPath = exportWithDetection(xmlContent.get(), baseFilename, "xml");
        if (xmlPath.isPresent()) {
          exportResults.put("LOIN-XML", xmlPath.get());
          System.out.println("   ✅ LOIN-XML exported: " + xmlPath.get());
        } else {
          System.out.println("   ❌ LOIN-XML export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ LOIN-XML export failed: No content received");
      }

      // IDS Export with auto-detection
      System.out.println("   🆔 Exporting as IDS...");
      Optional<byte[]> idsContent = client.exportLoinIds(selectedLoin.getGuid());
      if (idsContent.isPresent()) {
        String baseFilename = "loin_ids_" + selectedLoin.getGuid();
        Optional<Path> idsPath = exportWithDetection(idsContent.get(), baseFilename, "ids");
        if (idsPath.isPresent()) {
          exportResults.put("IDS", idsPath.get());
          System.out.println("   ✅ IDS exported: " + idsPath.get());
        } else {
          System.out.println("   ❌ IDS export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ IDS export failed: No content received");
      }

      // Summary with file type information
      System.out.println("\n📈 Export Summary: " + exportResults.size() + "/5 formats successful");
      for (Map.Entry<String, Path> entry : exportResults.entrySet()) {
        String filename = entry.getValue().getFileName().toString();
        String extension = filename.substring(filename.lastIndexOf('.') + 1).toUpperCase();
        System.out.println("   ✅ " + entry.getKey() + " (" + extension + "): " + entry.getValue());
      }

      if (exportResults.size() < 3) {
        System.out.println(
            "💡 Note: Some export formats may require special permissions or project setup");
        System.out.println(
            "💡 Content type detection helps ensure correct file extensions are used");
      }

    } catch (Exception e) {
      logger.error("❌ Error in LOIN export examples", e);
      System.err.println("❌ Error in LOIN export examples: " + e.getMessage());
    }
  }

  /** Main method to run LOIN export examples. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 BIM PORTAL LOIN EXPORT EXAMPLES");
    System.out.println("======================================================================");

    if (!checkCredentials()) {
      System.err.println("❌ Cannot run export examples without credentials");
      return;
    }

    if (!ExportUtils.isExportDirectoryWritable()) {
      System.err.println("❌ Export directory is not writable!");
      return;
    }

    System.out.println("🔧 Setting up authenticated client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      runLoinExportExamples(client);

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ LOIN EXPORT EXAMPLES COMPLETE!");
      System.out.println("======================================================================");
      System.out.println(
          "📂 Check the '" + BimPortalConfig.getExportPath() + "' directory for exported files");

      client.logout();
      System.out.println("👋 Logged out and session closed.");

    } catch (Exception e) {
      logger.error("❌ Error running LOIN export examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/export_examples/loin_export_example.py
"""
LOIN export examples for BIM Portal Python client.

This module demonstrates LOIN export workflows in multiple formats including 
PDF, OpenOffice, OKSTRA, LOIN-XML, and IDS with automatic content type detection.
"""

import os
import logging
from pathlib import Path
from typing import Dict, Optional

from dotenv import load_dotenv

from client.auth.auth_config import BIM_PORTAL_PASSWORD_ENV_VAR, BIM_PORTAL_USERNAME_ENV_VAR
from client.auth.auth_service_impl import AuthService
from client.enhanced_bim_client import EnhancedBimPortalClient
from client.config import BIMPortalConfig
from examples.utils.export_utils import ExportUtils

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()


def check_credentials() -> bool:
    """
    Check if credentials are available for authentication.
    
    Returns:
        True if credentials are configured
    """
    if not os.getenv(BIM_PORTAL_USERNAME_ENV_VAR) or not os.getenv(BIM_PORTAL_PASSWORD_ENV_VAR):
        print("=" * 60)
        print("WARNING: Credentials not found in environment variables.")
        print(f"Please set {BIM_PORTAL_USERNAME_ENV_VAR} and {BIM_PORTAL_PASSWORD_ENV_VAR} in .env file.")
        print("=" * 60)
        return False
    return True


def run_loin_export_examples(client: EnhancedBimPortalClient):
    """
    Run LOIN export examples with comprehensive format support.
    
    Args:
        client: Enhanced BIM Portal client
    """
    print("\n" + "=" * 60)
    print("📋 LOIN EXPORT EXAMPLES")
    print("=" * 60)
    
    print("\n1️⃣ Searching for available LOINs...")
    try:
        loins = client.search_loins()
        if not loins:
            print("🔭 No LOINs found for export")
            return
        
        print(f"✅ Found {len(loins)} LOINs:")
        for i, loin in enumerate(loins[:3], 1):
            print(f"   {i}. {loin.name} ({loin.guid})")
        
        selected_loin = loins[0]
        print(f"\n🎯 Using LOIN: '{selected_loin.name}'")
        
        export_results: Dict[str, Optional[Path]] = {}
        
        print("\n2️⃣ Exporting LOIN in multiple formats...")
        
        # PDF Export with auto-detection
        print("   📄 Exporting as PDF...")
        pdf_content = client.export_loin_pdf(selected_loin.guid)
        if pdf_content:
            base_filename = f"loin_pdf_{selected_loin.guid}"
            pdf_path = ExportUtils.export_with_detection(pdf_content, base_filename, "pdf")
            if pdf_path:
                export_results['PDF'] = pdf_path
                print(f"   ✅ PDF exported: {pdf_path}")
            else:
                print("   ❌ PDF export failed: Could not save file")
                export_results['PDF'] = None
        else:
            print("   ❌ PDF export failed: No content received")
            export_results['PDF'] = None
        
        # OpenOffice Export with auto-detection
        print("   📝 Exporting as OpenOffice...")
        odt_content = client.export_loin_openoffice(selected_loin.guid)
        if odt_content:
            base_filename = f"loin_odt_{selected_loin.guid}"
            odt_path = ExportUtils.export_with_detection(odt_content, base_filename, "odt")
            if odt_path:
                export_results['OpenOffice'] = odt_path
                print(f"   ✅ OpenOffice exported: {odt_path}")
            else:
                print("   ❌ OpenOffice export failed: Could not save file")
                export_results['OpenOffice'] = None
        else:
            print("   ❌ OpenOffice export failed: No content received")
            export_results['OpenOffice'] = None
        
        # OKSTRA Export with auto-detection
        print("   🗂️ Exporting as OKSTRA...")
        okstra_content = client.export_loin_okstra(selected_loin.guid)
        if okstra_content:
            base_filename = f"loin_okstra_{selected_loin.guid}"
            okstra_path = ExportUtils.export_with_detection(okstra_content, base_filename, "zip")
            if okstra_path:
                export_results['OKSTRA'] = okstra_path
                print(f"   ✅ OKSTRA exported: {okstra_path}")
            else:
                print("   ❌ OKSTRA export failed: Could not save file")
                export_results['OKSTRA'] = None
        else:
            print("   ❌ OKSTRA export failed: No content received")
            export_results['OKSTRA'] = None
        
        # LOIN-XML Export with auto-detection
        print("   🔗 Exporting as LOIN-XML...")
        xml_content = client.export_loin_xml(selected_loin.guid)
        if xml_content:
            base_filename = f"loin_xml_{selected_loin.guid}"
            xml_path = ExportUtils.export_with_detection(xml_content, base_filename, "xml")
            if xml_path:
                export_results['LOIN-XML'] = xml_path
                print(f"   ✅ LOIN-XML exported: {xml_path}")
            else:
                print("   ❌ LOIN-XML export failed: Could not save file")
                export_results['LOIN-XML'] = None
        else:
            print("   ❌ LOIN-XML export failed: No content received")
            export_results['LOIN-XML'] = None
        
        # IDS Export with auto-detection
        print("   🆔 Exporting as IDS...")
        ids_content = client.export_loin_ids(selected_loin.guid)
        if ids_content:
            base_filename = f"loin_ids_{selected_loin.guid}"
            ids_path = ExportUtils.export_with_detection(ids_content, base_filename, "ids")
            if ids_path:
                export_results['IDS'] = ids_path
                print(f"   ✅ IDS exported: {ids_path}")
            else:
                print("   ❌ IDS export failed: Could not save file")
                export_results['IDS'] = None
        else:
            print("   ❌ IDS export failed: No content received")
            export_results['IDS'] = None
        
        # Summary with file type information
        successful_exports = len([path for path in export_results.values() if path is not None])
        total_formats = len(export_results)
        
        print(f"\n📈 Export Summary: {successful_exports}/{total_formats} formats successful")
        
        for format_name, path in export_results.items():
            if path:
                extension = path.suffix[1:].upper() if path.suffix else "UNKNOWN"
                print(f"   ✅ {format_name} ({extension}): {path}")
            else:
                print(f"   ❌ {format_name}: Export failed")
        
        if successful_exports < 3:
            print("💡 Note: Some export formats may require special permissions or project setup")
            print("💡 Content type detection helps ensure correct file extensions are used")
        
        # Additional LOIN details
        print(f"\n📋 LOIN Details:")
        try:
            detailed_loin = client.get_loin(selected_loin.guid)
            if detailed_loin:
                print(f"   Name: {detailed_loin.name}")
                print(f"   GUID: {detailed_loin.guid}")
                if hasattr(detailed_loin, 'description') and detailed_loin.description:
                    print(f"   Description: {detailed_loin.description}")
                if hasattr(detailed_loin, 'version') and detailed_loin.version:
                    print(f"   Version: {detailed_loin.version}")
            else:
                print("   Could not retrieve detailed information")
        except Exception as e:
            print(f"   Error retrieving details: {e}")
        
    except Exception as e:
        logger.error(f"Error in LOIN export examples: {e}")
        print(f"❌ Error in LOIN export examples: {e}")


def main():
    """Main method to run LOIN export examples."""
    print("=" * 70)
    print("🚀 BIM PORTAL LOIN EXPORT EXAMPLES")
    print("=" * 70)
    
    if not check_credentials():
        print("❌ Cannot run export examples without credentials")
        return
    
    if not ExportUtils.is_export_directory_writable():
        print("❌ Export directory is not writable!")
        return
    
    print("🔧 Setting up authenticated client...")
    
    try:
        auth_service = AuthService()
        client = EnhancedBimPortalClient(
            auth_service=auth_service, 
            base_url=BIMPortalConfig.BASE_URL
        )
        
        # Run LOIN export examples
        run_loin_export_examples(client)
        
        print("\n" + "=" * 70)
        print("✅ LOIN EXPORT EXAMPLES COMPLETE!")
        print("=" * 70)
        print(f"📂 Check the '{BIMPortalConfig.EXPORT_DIRECTORY}' directory for exported files")
        
    except Exception as e:
        logger.error(f"Error running LOIN export examples: {e}")
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()
